domain/application/state/charm.go

```diff
@@ -594,46 +595,159 @@ func (s *CharmState) GetCharm(ctx context.Context, id corecharm.ID) (charm.Charm
 func (s *CharmState) SetCharm(ctx context.Context, charm charm.Charm, charmArgs charm.SetStateArgs) (corecharm.ID, error) {
 	db, err := s.DB()
 	if err != nil {
-		return "", errors.Trace(err)
+		return "", internalerrors.Capture(err)
 	}
 
 	id, err := corecharm.NewID()
 	if err != nil {
-		return "", fmt.Errorf("failed to set charm: %w", err)
+		return "", internalerrors.Errorf("setting charm: %w", err)
 	}
 
 	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		// Check the charm doesn't already exist, if it does, return an already
 		// exists error. Also doing this early, prevents the moving straight
 		// to a write transaction.
 		if _, err := s.checkCharmReferenceExists(ctx, tx, charmArgs.ReferenceName, charmArgs.Revision); err != nil {
-			return errors.Trace(err)
+			return internalerrors.Capture(err)
 		}
 
 		if err := s.setCharm(ctx, tx, id, charm, charmArgs.ArchivePath); err != nil {
-			return errors.Trace(err)
+			return internalerrors.Capture(err)
 		}
 
 		if err := s.setCharmHash(ctx, tx, id, charmArgs.Hash); err != nil {
-			return errors.Trace(err)
+			return internalerrors.Capture(err)
 		}
 
 		if err := s.setCharmInitialOrigin(ctx, tx, id, charmArgs.ReferenceName, charmArgs.Source, charmArgs.Revision, charmArgs.Version); err != nil {
-			return errors.Trace(err)
+			return internalerrors.Capture(err)
 		}
 
 		if err := s.setCharmPlatform(ctx, tx, id, charmArgs.Platform); err != nil {
-			return errors.Trace(err)
+			return internalerrors.Capture(err)
 		}
 
 		return nil
 	}); err != nil {
-		return "", fmt.Errorf("failed to set charm: %w", err)
+		return "", internalerrors.Errorf("setting charm: %w", err)
 	}
 
 	return id, nil
 }
 
+// ListCharmsWithOrigin returns a list of charms with the specified origin.
+// We require the origin, so we can reconstruct the charm URL for the
+// client response.
+func (s *CharmState) ListCharmsWithOrigin(ctx context.Context) ([]charm.CharmWithOrigin, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, internalerrors.Capture(err)
+	}
+
+	query := `
+SELECT cm.uuid AS &charmNameWithOrigin.uuid,
+	cm.name AS &charmNameWithOrigin.name,
+	co.reference_name AS &charmNameWithOrigin.reference_name,
+	co.source AS &charmNameWithOrigin.source,
+	co.revision AS &charmNameWithOrigin.revision,
+	cp.architecture_id AS &charmNameWithOrigin.architecture_id
+FROM v_charm_metadata AS cm
+JOIN v_charm_origin AS co ON cm.uuid = co.charm_uuid
+JOIN charm_platform AS cp ON cm.uuid = cp.charm_uuid;
+`
+	stmt, err := s.Prepare(query, charmID{}, charmNameWithOrigin{})
```

Comment:
`charmID` isn't referred to in your query.

@Aflynn50 This should probably cause SQLair `.Prepare` to fail imo

Reply:
Yeah, that would be a good idea.

Reply:
Ah, this is already fixed https://github.com/canonical/sqlair/pull/153

Reply:
I guess I forgot to update the SQLair version in juju, try doing:
```
go get -u github.com/canonical/sqlair
```
If theres a ton of places you have to fix this then I don't mind doing it, but if its easy might as well just do it here.